server/src-lib/Hasura/GraphQL/Execute.hs

```diff
@@ -486,76 +466,57 @@ insertBatchResults remoteValue batch hasuraResp =
-                        ( batchRelationshipKeyIndex batch
-                        , batchRelationshipKeyToMake batch)
-                        (peelOffNestedFields
-                           (batchNestedFields batch)
-                           (snd (theRemoteHash)))
-                        hasuraHash)
-                     (batchPhantoms batch))
-                , remainingRemoteHashes)
-    inHashmap (RelFieldPath ((idx, G.Alias (G.Name key)) Seq.:<| rest)) hasuraHash remoteHash =
-      case OJ.lookup key hasuraHash of
-        Nothing ->
-          Left
-            ("Couldn't find expected key " <> show key <> " in " <>
-             show (OJ.toEncJSON (OJ.Object hasuraHash)) <>
-             ", while traversing " <>
-             show (OJ.toEncJSON (OJ.Object hasuraHash0)))
-        Just currentValue ->
-          fmap
-            (\(newValue, remoteCandidates) ->
-               (OJ.insert (idx, key) newValue hasuraHash, remoteCandidates))
-            (inValue rest currentValue remoteHash)
+    inHashmap (RelFieldPath p) hasuraHash remoteHash = case p of
+      Seq.Empty ->
+        case remoteHash of
+          [] -> Left $ err <> show (OJ.toEncJSON $ OJ.Object hasuraHash)
+            where err = case cardinality of 
+                          One -> "Expected one remote object but got none, while traversing " 
+                          Many -> "Expected many objects but got none, while traversing " 
+
+          (theRemoteHash:remainingRemoteHashes) 
+            | cardinality == One && not (null remainingRemoteHashes) ->
+                Left $
+                  "Expected one remote object but got many, while traversing " <> showHash hasuraHash
+            | otherwise -> 
+                Right
+                  ( (foldl'
+                       (flip OJ.delete)
+                       (OJ.insert
+                          (batchRelationshipKeyIndex, batchRelationshipKeyToMake)
+                          (peelOffNestedFields batchNestedFields (snd theRemoteHash))
+                          hasuraHash)
+                       batchPhantoms)
+                  , remainingRemoteHashes)
+ 
+      ((idx, G.Alias (G.Name key)) Seq.:<| rest) ->
+        case OJ.lookup key hasuraHash of
+          Nothing ->
+            Left $ 
+              "Couldn't find expected key " <> show key <> " in " <> showHash hasuraHash <>
+              ", while traversing " <> showHash hasuraHash0
+          Just currentValue ->
+            first (\newValue -> OJ.insert (idx, key) newValue hasuraHash)
+              <$> inValue rest currentValue remoteHash
+      where
+        showHash = show . OJ.toEncJSON . OJ.Object
+
+    -- TODO what is this supposed to do, and what does the return values mean?
     inValue ::
          Seq.Seq (Int, G.Alias)
       -> OJ.Value
       -> [(Text, OJ.Value)]
       -> Either String (OJ.Value, [(Text, OJ.Value)])
     inValue path currentValue remoteHash =
+      -- TODO refactor with MonadError
```

Comment:
The problem with MonadError is it short-circuits exeuction rather than accumulate all errors. We should have a MonadGQError or something that can accumulate errors when thrown.

Reply:
Here I'm just referring to using `MonadError` to avoid the nested Either cases marching off the left-hand side of page, rather than the error accumulating machinery. But I'll have to just try and see how it looks.